lib/datadog/statsd/sender.rb

```diff
@@ -2,20 +2,34 @@
 
 module Datadog
   class Statsd
+    # Sender is using a companion thread to flush and pack messages
+    # in a `MessageBuffer`.
+    # The communication with this thread is done using a `Queue`.
+    # If the thread is dead, it is starting a new one to avoid having a blocked
+    # Sender with no companion thread to communicate with (most of the time, having
+    # a dead companion thread means that a fork just happened and that we are
+    # running in the child process).
     class Sender
       CLOSEABLE_QUEUES = Queue.instance_methods.include?(:close)
 
-      def initialize(message_buffer)
+      def initialize(message_buffer, logger: nil)
         @message_buffer = message_buffer
+        @logger = logger
+        @mx = Mutex.new
       end
 
       def flush(sync: false)
-        # don't try to flush if there is no message_queue instantiated
-        return unless message_queue
-
-        message_queue.push(:flush)
+        # don't try to flush if there is no message_queue instantiated or
+        # no companion thread running
+        if !message_queue || !sender_thread.alive?
+          @logger.warn { "Statsd: can't flush: no message queue or sender_thread alive" } if @logger
+          return
+        end
 
-        rendez_vous if sync
+        @mx.synchronize {
+          message_queue.push(:flush)
+          rendez_vous if sync
+        }
       end
 
       def rendez_vous
```

Comment:
Since `#rendez_vous` is public, should it also perform similar checks to `#flush` and use the lock as well?